pylint/checkers/variables.py

```diff
@@ -743,10 +743,109 @@ def _uncertain_nodes_in_except_blocks(
                     # if one of the except blocks does not define the name in question,
                     # raise, or return. See: https://github.com/PyCQA/pylint/issues/5524.
                     continue
+
+            if NamesConsumer._check_loop_finishes_via_except(
+                node, other_node_statement.parent.parent
+            ):
+                continue
+
             # Passed all tests for uncertain execution
             uncertain_nodes.append(other_node)
         return uncertain_nodes
 
+    @staticmethod
+    def _check_loop_finishes_via_except(
+        node: nodes.NodeNG, other_node_try_except: nodes.TryExcept
+    ) -> bool:
+        """Check for a case described in https://github.com/PyCQA/pylint/issues/5683.
+        It consists of a specific control flow scenario where the only
+        non-break exit from a loop consists of the very except handler we are
+        examining, such that code in the `else` branch of the loop can depend on it
+        being assigned.
+
+        Example:
+
+        for _ in range(3):
+            try:
+                raise Exception
+            except:
+                name = 1  <-- only non-break exit from loop
+            else:
+                break
+        else:
+            print(name)
+        """
+        if not other_node_try_except.orelse:
+            return False
+        closest_loop: Optional[
+            Union[nodes.For, nodes.While]
+        ] = utils.get_node_first_ancestor_of_type(
+            node, (nodes.For, nodes.While)  # type: ignore
```

Comment:
Can you add the error this is ignoring? `mypy` has not yet implemented the feature, but we're trying not to do general type ignores.

Btw, isn't this returning the correct typing? DO we need to update `get_node_first_ancestor_of_type`?

Reply:
I agree typing ``get_node_first_ancestor_of_type`` in a prior MR would probably make sense here. But it seems it's node_ancestors from astroid's ``NodeNG`` that should be typed ?
```python
def get_node_first_ancestor_of_type(
    node: nodes.NodeNG, ancestor_type: Union[Type[T_Node], Tuple[Type[T_Node]]]
) -> Optional[T_Node]:
    """Return the first parent node that is any of the provided types (or None)"""
    for ancestor in node.node_ancestors():
        if isinstance(ancestor, ancestor_type):
            return ancestor
    return None
```

Reply:
Ah, it needs a `...`! I'll fix in another PR (and the similar function just below it)

Reply:
Ah, that's a nasty oversight from me in the review. Good catch!